presto-main/src/main/java/com/facebook/presto/operator/DistinctLimitOperator.java

```diff
@@ -186,16 +203,50 @@ public void addInput(Page page)
                 }
             }
         }
+        groupByIds = null;
+        inputPage = null;
+
+        Page result = null;
         if (!pageBuilder.isEmpty()) {
-            outputPage = pageBuilder.build();
+            result = pageBuilder.build();
+            pageBuilder.reset();
         }
+
+        updateMemoryReservation();
+        return result;
     }
 
-    @Override
-    public Page getOutput()
+    private boolean processUnfinishedWork()
     {
-        Page result = outputPage;
-        outputPage = null;
-        return result;
+        verify(unfinishedWork != null);
+        if (!unfinishedWork.process()) {
+            return false;
+        }
+        groupByIds = unfinishedWork.getResult();
+        unfinishedWork = null;
+        return true;
+    }
+
+    /**
+     * Update memory usage with extra memory needed.
```

Comment:
nit: What's `extra memory` here? We update the memory reservation here so we may remove this line.